[[ Bug 15924 ]] Add javascript implementation of liburl to the emscripten engine.

livecodemichael opened the pull request:
A liburl driver has been added that implements the liburl handlers for http(s) urls. The driver calls javascript functions (via do as javascript) to create xml http requests for each url fetched by liburl.

Relies on https://github.com/livecode/livecode-ide/pull/1705

livecodemichael committed:
[[ Bug 15924 ]] Add javascript version of liburl to the emscripten engine.

A liburl driver has been added that implements the liburl handlers for http(s)
urls. The driver calls javascript functions (via do as javascript) to create
xml http requests for each url fetched by liburl.

livecodemichael committed:
[[ Bug 15924 ]] Make IDE inclusion dependencies take platform into account.

livecodemichael committed:
[[ Bug 15924 ]] Implement emscripten auxiliary stack support.

The the list of auxiliary stacks is now based on the script libraries standalone
setting. The order of loading is managed by a simple manifest file.

livecodemichael committed:
[[ Bug 15924 ]] Add emscripten liburl to ide.

livecodemichael committed:
[[ Bug 15924 ]] Fix whitespace changes.

livecodeian commented on Installer/package.txt:
```diff
@@ -833,6 +833,7 @@ component Toolset
 		stack ide-support:revsaveasiosstandalone.livecodescript
 		stack ide-support:revsaveasstandalone.livecodescript
 		stack ide-support:revsblibrary.livecodescript
+		stack ide-support:revhtml5liburllibrary.livecodescript
```
looks like this should be revhtml5urllibrary.livecodescript

livecodeian commented on engine/src/object.h:
```diff
@@ -896,7 +896,10 @@ class MCObject :
+            return parent;
+        else
+            return NULL;
 	}
```
How does this differ from the current behaviour?

livecodemichael replied:
Yeah, I was meaning to ask how best to handle this.

The parent is an object handle, which will trigger and assertion failure when trying to fetch a pointer to the original object if the original object is not valid.

In the case of setting the home stack during startup (MCDispatch::changehome), the parent of the new home stack is not set, which triggers the assertion failure.

livecodeian commented on ide-support/revhtml5urllibrary.livecodescript:
```diff
@@ -0,0 +1,452 @@
+      if tParam is not a number and tParam is not among the words of "true false" then
+         replace quote with slash & quote in tParam
+         put quote & tParam & quote into tParam
+      end if
```
This doesn't look quite right - should that be backslash? We may also need to escape other chars here - CR, LF, \, etc.

livecodemichael replied:
Yes, this was put in quickly with the intention of going back to fix later. Good spot!

livecodeian replied:
You need to replace backslashes first, otherwise you'll be escaping them in the escape sequences - e.g. an escaped return will end up as '\\n' rather than just '\n'

livecodemichael committed:
[[ Bug 15924 ]] Fix a couple of typos

livecodemichael committed:
[[ Bug 15924 ]] Add docs for HTML5 URL syntax.

livecodemichael committed:
[[ Bug 15924 ]] Make sure we use the correct escape sequences when passing string to javascript functions.

livecodemichael committed:
[[ Bug 15924 ]] Fix crash when fetching font atss during HTML5 startup.

The default stack can be null during the HTML5 engine startup
(i.e. when loading aux stacks). Make sure we don't try to fetch
the font atts of the default stack when this is the case.

livecodeian commented on ide-support/revhtml5urllibrary.livecodescript:
```diff
@@ -0,0 +1,455 @@
+         replace quote with backslash & quote in tParam
+         replace return with "\n" in tParam
+         replace cr with "\r" in tParam
+         replace backslash with "\\" in tParam
```
I commented on the commit but it doesn't appear in the PR:
You need to replace backslashes first, otherwise you'll be escaping them in the escape sequences - e.g. an escaped return will end up as '\\\\n' rather than just '\\n'

livecodemichael committed:
[[ Bug 15924 ]] Escape backslashes first when passing params to javascript.

livecodeian commented:
@livecode-vulcan review ok 030cc96

livecode-vulcan commented:
💙 review by @livecodeian ok 030cc96b2cc48d06cf02e7acdc975f0620ac6e66

livecode-vulcan commented:
😎 test success 030cc96b2cc48d06cf02e7acdc975f0620ac6e66

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/603)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/619)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/593)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/536)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/539)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/597)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/617)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/602)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/619)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/601)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/612)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/634)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/598)
